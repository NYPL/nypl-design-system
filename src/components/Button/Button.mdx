import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";
import Link from "../Link/Link";

<Meta of={ButtonStories} />

# Button

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.0.4`      |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#button-groups" target="_self">Button Groups</Link>}
- {<Link href="#with-icons" target="_self">With Icons</Link>}
- {<Link href="#patterns" target="_self">Patterns</Link>}
- {<Link href="#other-examples" target="_self">Other Examples</Link>}
- {<Link href="#sizes" target="_self">Sizes</Link>}

## Overview

<Description of={ButtonStories} />

Update the `buttonType` in the Controls section to see the types of buttons
that can be rendered.

- `"primary"` is used for actions that move the user forward. This is the default.
- `"secondary"` is used for actions that move the user back, such as cancellations.
- `"text"` is used for equally weighted actions in a text based list.
- `"callout"` is used for call to action text such as "Donate".
- `"noBrand"` is a variant used when there is no brand and will display the
  background color as black.
- `"pill"` is used for equally weighted actions within cards and always in a set.
- `"link"` has been deprecated and should not be used.

When one and only one `Icon` component is passed inside a `Button` component with
no text, it will automatically be configured to use the `"iconOnly"` type.

## Component Props

<Canvas of={ButtonStories.WithControls} />

<ArgTypes of={ButtonStories.WithControls} />

## Accessibility

The Reservoir `Button` component renders an HTML `<button>` element that is
accessible. `Button`s should be used to perform user actions or submit forms,
never to link and go to another page. For `Button`s without text and only an
icon, make sure to pass an `aria-label` value to the `Button`. `Button`s cannot
have images of text.

`Button`s can be rendered in three sizes through the `size` prop: `"small"`,
`"medium"`, and `"large"`. The default size is `"medium"`. Even though the `"small"`
size is available, we recommend to use the `"medium"` and `"large"` sizes.

The `Button` component also has a `screenreaderOnlyText` prop that can be used to
add additional text to the component that is only visible to screen readers. This
can be used to provide additional context to the `Button` when the text is short,
such as visible "Read more..." text.

Resources:

- [W3C WAI ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button)
- [W3C WAI ARIA Button Examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html)
- [MDN ARIA: button role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role)
- [Chakra UI Button](https://chakra-ui.com/docs/components/form/button)
- [W3C WAI WCAG Target Size](https://www.w3.org/WAI/WCAG21/Understanding/target-size.html)

## Button Groups

Primary and secondary buttons should be grouped like in the following example.
The `ButtonGroup` component should be used to wrap the `Button` component.

<Canvas of={ButtonStories.ButtonGroups} />

## With Icons

The `Icon` component collaborates great with the `Button` component. Check out
[`src/components/Icons/Icon.tsx`](https://github.com/NYPL/nypl-design-system/blob/development/src/components/Icons/Icon.tsx)
to see the full list of icon names that can be rendered.

The icon can be placed to the left or the right of the button text.

<Canvas of={ButtonStories.TextWithIcon} />

Text in a button is optional if an icon is rendered, but make sure that there
is an `aria-label` passed to the `Button` component `aria-label="Previous"`.

<Canvas of={ButtonStories.IconOnly} />

## Patterns

Buttons are blue by default but should be red when it is intended to be used as
a call out. The "Back to Top" button must be of `buttonType="secondary"` and
must include an up arrow icon on the right side.

<Canvas of={ButtonStories.Patterns} />

## Other Examples

The variations modified by the `buttonType` prop:

<Canvas of={ButtonStories.ButtonTypes} />

Modifying the `isDisabled` prop:

<Canvas of={ButtonStories.DisabledButtons} />

## Sizes

Update the size of the `Button` component with the `size` prop and values:
`"small"`, `"medium"`, `"large"`. The default size is `"medium"`.

<Canvas of={ButtonStories.ButtonSizes} />
