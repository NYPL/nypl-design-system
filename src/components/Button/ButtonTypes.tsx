export enum ButtonTypes {
  Primary = "filled",
  Secondary = "outline",
  Pill = "pill",
  Link = "link",
}
