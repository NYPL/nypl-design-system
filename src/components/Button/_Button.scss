@mixin button-base {
  @include space-inset-xs;

  border-radius: 2px;
  cursor: pointer;
  display: flex;
  justify-content: center;
  text-decoration: none;

  .icon {
    @include icon;

    fill: currentColor;

    &--icon-left {
      margin-right: var(--space-xs);
    }

    &--icon-right {
      margin-left: var(--space-xs);
    }
  }
}

@mixin button--filled {
  @include button-base;

  background-color: var(--ui-link-primary);
  border: none;
  color: var(--ui-white);

  &:visited {
    color: var(--ui-white);
  }

  &:hover {
    background-color: var(--ui-link-secondary);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--outline {
  @include button-base;

  background-color: var(--ui-white);
  border: 1px solid var(--ui-gray-light);
  color: inherit;

  &:visited {
    color: inherit;
  }

  &:hover {
    background-color: var(--ui-gray-xxlight);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--link {
  background-color: transparent;
  color: var(--ui-link-primary);
  text-decoration: underline;

  &--no-underline {
    text-decoration: none;
  }

  &:hover {
    background-color: transparent;
    color: var(--ui-link-secondary);
  }

  &:disabled {
    background-color: transparent;
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--pill {
  @include button-base;
  @include button--outline;

  border-radius: 20px;
}

.button {
  @include button--filled;

  &--link {
    @include button--link;
  }

  &--outline {
    @include button--outline;
  }

  &--pill {
    @include button--pill;
  }
}
