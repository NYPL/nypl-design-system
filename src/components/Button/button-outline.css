@import "../Icons/_Icons.scss";

@mixin button-base {
  border-radius: 2px;
  cursor: pointer;
  display: flex;
  justify-content: center;
  padding: var(--space-xs);
  text-decoration: none;

  .icon {
    @include icon;

    fill: currentColor;

    &--icon-left {
      margin-right: var(--space-xs);
    }

    &--icon-right {
      margin-left: var(--space-xs);
    }
  }
}

@mixin button--outline {
  @include button-base;

  background-color: var(--ui-white);
  border: 1px solid var(--ui-gray-light);
  color: inherit;

  &:visited {
    color: inherit;
  }

  &:hover {
    background-color: var(--ui-gray-xxlight);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
  }
}
