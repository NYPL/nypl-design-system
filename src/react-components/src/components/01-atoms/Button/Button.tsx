import * as React from "react";
import Icon from "../Images/Icons/Icon";
import bem from "../../../utils/bem";

export interface ButtonProps {
  id?: string;
  /** The action to perform on the <button>'s onClick function */
  callback: (event: React.MouseEvent) => void;
  content: string | JSX.Element;
  attributes?: {};
  modifiers?: string[];
  blockName?: string;
  large?: boolean;
  type?: string;
  mouseDown?: boolean;

  iconPosition?: string;
  iconName?: string;
  iconDecorative?: boolean;
  iconRole?: string;
}

export default class Button extends React.Component<ButtonProps, {}> {
  static defaultProps = {
    mouseDown: false
  };

  constructor(props: ButtonProps) {
    super(props);
  }

  render(): JSX.Element {
    const { id, callback, content, attributes, modifiers, blockName, type, mouseDown, iconPosition, iconName, iconDecorative, iconRole } = this.props;
    if (type) {
      if (!(type === "outline" || type === "filled")) {
        throw new Error("Type can only be 'outline' or 'filled'");
      }
    }

    let buttonModifiers = modifiers ? modifiers : [];
    if (!buttonModifiers.find((mod) => mod === type)) {
      buttonModifiers.push(type);
    }

    let btnContent = [content];
    let button_base_class = "button";

    let iconProps = {
      name: iconName,
      blockName: button_base_class,
      modifiers: ["small"],
      decorative: iconDecorative,
      desc: iconDecorative,
      role: iconDecorative ? undefined : iconRole,
      title: iconDecorative
    };

    if (iconPosition) {
      buttonModifiers.push("icon");

      if (iconPosition === "left") {
        buttonModifiers.push("icon-left");
        iconProps.modifiers.push("icon-left");
      } else if (iconPosition === "right") {
        buttonModifiers.push("icon-right");
        iconProps.modifiers.push("icon-right");
      }

      btnContent.push(React.createElement(
        Icon,
        iconProps
      ));
    }

    let btnProps = {
      className: bem(button_base_class, buttonModifiers, blockName),
      type: "submit"
    };
    if (id) {
      btnProps[id] = id;
    }
    let btnCallback = mouseDown ? { onMouseDown: callback } : { onClick: callback };

    return React.createElement(
      "button",
      { ...btnProps, ...btnCallback, ...attributes },
      btnContent
    );
  }
}
