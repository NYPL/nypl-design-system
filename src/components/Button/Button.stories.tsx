import { VStack } from "@chakra-ui/react";
import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";

import Button from "./Button";
import ButtonGroup from "../ButtonGroup/ButtonGroup";
import Icon from "../Icons/Icon";
import Table from "../Table/Table";

const iconNames = [
  "accessibilityFull",
  "accessibilityPartial",
  "actionCheckCircle",
  "actionCheckCircleFilled",
  "actionExit",
  "actionHelpDefault",
  "actionHelpOutline",
  "alertNotificationImportant",
  "actionLaunch",
  "actionRegistration",
  "actionPower",
  "actionSettings",
  "alertNotificationImportant",
  "alertWarningFilled",
  "alertWarningOutline",
  "arrow",
  "building",
  "check",
  "clock",
  "close",
  "decorativeEnvelope",
  "decorativeLibraryCard",
  "decorativeShoppingBag",
  "download",
  "errorFilled",
  "errorOutline",
  "fileTypeAudio",
  "fileTypeDoc",
  "fileTypeGenericDoc",
  "fileTypeImage",
  "fileTypePdf",
  "fileTypeSpreadsheet",
  "fileTypeVideo",
  "headset",
  "legacyAccountFilled",
  "legacyAccountUnfilled",
  "legacySocialFacebook",
  "legacySocialInstagram",
  "legacySocialTwitter",
  "legacySocialYoutube",
  "locator",
  "minus",
  "plus",
  "search",
  "socialFacebook",
  "socialInstagram",
  "socialTumblr",
  "socialTwitter",
  "socialYoutube",
  "speakerNotes",
  "utilityAccountFilled",
  "utilityAccountUnfilled",
  "utilityHamburger",
  "utilitySearch",
];

const meta: Meta<typeof Button> = {
  title: "Components/Form Elements/Button",
  component: Button,
  decorators: [withDesign],
  argTypes: {
    buttonText: { description: "Only used for Storybook" },
    buttonType: {
      table: { defaultValue: { summary: "primary" } },
    },
    className: { control: false },
    displayIcon: { description: "Only used for Storybook." },
    displayIconLeft: {
      description: "Move the icon left or right. Only used for Storybook.",
    },
    iconType: {
      control: { type: "select" },
      options: iconNames,
      description: "Select the icon. Only used for Storybook.",
    },
    id: { control: false },
    isDisabled: { table: { defaultValue: { summary: false } } },
    mouseDown: { table: { defaultValue: { summary: false } } },
    onClick: { control: false },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

/**
 * Main Story for the Button component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    buttonText: "Button Text",
    buttonType: "primary",
    className: undefined,
    displayIcon: false,
    displayIconLeft: true,
    iconType: "search",
    id: "button-id",
    isDisabled: false,
    mouseDown: undefined,
    onClick: undefined,
    size: "medium",
    type: "button",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11477%3A2298",
    },
    jest: ["Button.test.tsx"],
  },
  render: () => <Button id="button-id">Button Text</Button>,
};

// The following are additional Button example Stories.
export const ButtonGroups: Story = {
  render: () => (
    <ButtonGroup>
      <Button buttonType="secondary" id="group-1">
        Button
      </Button>
      <Button id="group-2">Submit</Button>
    </ButtonGroup>
  ),
};
export const WithIcons: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <VStack align="left" spacing="input.group.button.vstack" width="150px">
        <Button id="icon-1">
          <Icon name="search" align="left" size="small" />
          Button Text
        </Button>
        <Button id="icon-2">
          Button Text
          <Icon name="search" align="right" size="small" />
        </Button>
      </VStack>
      <ButtonGroup>
        <Button buttonType="secondary" id="icon-left">
          <Icon
            name="arrow"
            iconRotation="rotate90"
            align="left"
            size="small"
          />
          Previous
        </Button>
        <Button buttonType="secondary" id="icon-right">
          Next
          <Icon
            name="arrow"
            iconRotation="rotate270"
            align="right"
            size="small"
          />
        </Button>
      </ButtonGroup>
    </VStack>
  ),
};
export const IconOnly: Story = {
  render: () => (
    <ButtonGroup>
      <Button aria-label="Previous" buttonType="secondary" id="prev-btn">
        <Icon name="arrow" iconRotation="rotate90" size="small" />
      </Button>
      <Button aria-label="Next" buttonType="secondary" id="next-btn">
        <Icon name="arrow" iconRotation="rotate270" size="small" />
      </Button>
      <Button aria-label="Close" buttonType="secondary" id="close-btn">
        <Icon name="close" size="small" />
      </Button>
    </ButtonGroup>
  ),
};
export const Patterns: Story = {
  render: () => (
    <ButtonGroup>
      <Button buttonType="callout" id="donate-btn">
        Donate to this library
      </Button>
      <Button buttonType="secondary" id="top-btn">
        Back to Top
        <Icon
          name="arrow"
          iconRotation="rotate180"
          size="small"
          className="right"
          align="right"
        />
      </Button>
    </ButtonGroup>
  ),
};
export const ButtonTypes: Story = {
  render: () => (
    <ButtonGroup alignItems="center">
      <Button id="primary-btn">Primary</Button>
      <Button buttonType="secondary" id="secondary-btn">
        Secondary
      </Button>
      <Button buttonType="text" id="text-btn">
        Text
      </Button>
      <Button buttonType="callout" id="callout-btn">
        Callout
      </Button>
      <Button buttonType="noBrand" id="nobrand-btn">
        No Brand
      </Button>
      <Button buttonType="pill" id="pill-btn">
        Pill
      </Button>
      <Button buttonType="link" id="link-btn">
        Link (deprecated)
      </Button>
    </ButtonGroup>
  ),
};
export const DisabledButtons: Story = {
  render: () => (
    <ButtonGroup>
      <Button id="enabled-btn" isDisabled={false}>
        Enabled
      </Button>
      <Button id="disabled-btn" isDisabled>
        isDisabled
      </Button>
    </ButtonGroup>
  ),
};

const sizeHeaders = ["--", "small", "medium (default)", "large"];
const buttonData = [
  [
    "primary",
    <Button id="p-small" key="p-small" size="small">
      Primary
    </Button>,
    <Button id="p-medium" key="p-medium" size="medium">
      Primary
    </Button>,
    <Button id="p-large" key="p-large" size="large">
      Primary
    </Button>,
  ],
  [
    "secondary",
    <Button buttonType="secondary" id="s-small" key="s-small" size="small">
      Secondary
    </Button>,
    <Button buttonType="secondary" id="s-medium" key="s-medium" size="medium">
      Secondary
    </Button>,
    <Button buttonType="secondary" id="s-large" key="s-large" size="large">
      Secondary
    </Button>,
  ],
  [
    "text",
    <Button buttonType="text" id="t-small" key="t-small" size="small">
      Text
    </Button>,
    <Button buttonType="text" id="t-medium" key="t-medium" size="medium">
      Text
    </Button>,
    <Button buttonType="text" id="t-large" key="t-large" size="large">
      Text
    </Button>,
  ],
  [
    "callout",
    <Button buttonType="callout" id="c-small" key="c-small" size="small">
      Callout
    </Button>,
    <Button buttonType="callout" id="c-medium" key="c-medium" size="medium">
      Callout
    </Button>,
    <Button buttonType="callout" id="c-large" key="c-large" size="large">
      Callout
    </Button>,
  ],
  [
    "noBrand",
    <Button buttonType="noBrand" id="nb-small" key="nb-small" size="small">
      No Brand
    </Button>,
    <Button buttonType="noBrand" id="nb-medium" key="nb-medium" size="medium">
      No Brand
    </Button>,
    <Button buttonType="noBrand" id="nb-large" key="nb-large" size="large">
      No Brand
    </Button>,
  ],
  [
    "pill",
    <Button buttonType="pill" id="pill-small" key="pill-small" size="small">
      Pill
    </Button>,
    <Button buttonType="pill" id="pill-medium" key="pill-medium" size="medium">
      Pill
    </Button>,
    <Button buttonType="pill" id="pill-large" key="pill-large" size="large">
      Pill
    </Button>,
  ],
  [
    "link (deprecated)",
    <Button buttonType="link" id="l-small" key="l-small" size="small">
      Link
    </Button>,
    <Button buttonType="link" id="l-medium" key="l-medium" size="medium">
      Link
    </Button>,
    <Button buttonType="link" id="l-large" key="l-large" size="large">
      Link
    </Button>,
  ],
];

export const ButtonSizes: Story = {
  render: () => (
    <Table
      columnHeaders={sizeHeaders}
      id="all-sizes"
      showRowDividers
      tableData={buttonData}
      useRowHeaders
      sx={{
        tbody: {
          th: {
            verticalAlign: "middle",
          },
          td: {
            verticalAlign: "middle",
          },
        },
      }}
    />
  ),
};
