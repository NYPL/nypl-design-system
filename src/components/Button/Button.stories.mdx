import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";

import Button from "./Button";
import { ButtonTypes } from "./ButtonTypes";
import Icon from "../Icons/Icon";
import { IconNames, IconRotationTypes } from "../Icons/IconTypes";
import { action } from "@storybook/addon-actions";
import getCSSVariable from "../../helpers/getCSSVariable";
import { getCategory } from "../../utils/componentCategories";

<Meta
  title={getCategory("Button")}
  component={Button}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11477%3A2298",
    },
  }}
/>

# Button

<Description of={Button} />

Update the `buttonType` in the Controls section to see the types of buttons
that can be rendered.

- `ButtonTypes.Primary` is used for actions that move the user forward.
- `ButtonTypes.Secondary` is used for actions taht move the user back, such as cancellations
- `ButtonTypes.Pill` is used for equally weighted actions within cards and always in a set

<Preview withToolbar>
  <Story name="Basic">{args => <Button {...args}>Button Text</Button>}</Story>
</Preview>

<ArgsTable story="Basic" />

## Button Groups

Primary and secondary buttons can be used in groups like so:

<Canvas>
  <Story name="Button Groups">
    <div
      style={{
        display: "flex",
        justifyContent: "space-between",
        width: "150px",
      }}
    >
      <Button buttonType={ButtonTypes.Secondary}>Button</Button>
      <Button buttonType={ButtonTypes.Primary}>Submit</Button>
    </div>
    <br />
    <div
      style={{
        display: "flex",
        justifyContent: "space-between",
        width: "250px",
      }}
    >
      <Button buttonType={ButtonTypes.Secondary}>
        <span style={{ width: "100px" }}>Button</span>
      </Button>
      <Button buttonType={ButtonTypes.Primary}>
        <span style={{ width: "100px" }}>Submit</span>
      </Button>
    </div>
  </Story>
</Canvas>

## With Icons

The `Icon` component collaborates great with the `Button` component. Check out
`src/components/Icons/IconTypes` to see the list of icon and logo names that
can be rendered or check out the Controls section for this story.

<Canvas>
  <Story
    name="With Icon"
    args={{
      displayLeft: true,
      buttonText: "Button Text",
      iconType: IconNames.search,
    }}
    argTypes={{
      iconType: {
        control: {
          type: "select",
          options: IconNames,
        },
      },
    }}
  >
    {args => (
      <Button {...args}>
        {args.displayLeft && (
          <Icon
            name={args.iconType}
            decorative={true}
            modifiers={["small", "icon-left"]}
          />
        )}
        {args.buttonText}
        {!args.displayLeft && (
          <Icon
            name={args.iconType}
            decorative={true}
            modifiers={["small", "icon-right"]}
          />
        )}
      </Button>
    )}
  </Story>
</Canvas>

The icon can be placed to the left or the right of the button text.

<Canvas>
  <div
    style={{
      display: "flex",
      justifyContent: "space-between",
      width: "200px",
    }}
  >
    <Button buttonType={ButtonTypes.Secondary}>
      <Icon
        name={IconNames.arrow}
        decorative={true}
        iconRotation={IconRotationTypes.rotate90}
        modifiers={["small", "icon-left"]}
      />
      Previous
    </Button>
    <Button buttonType={ButtonTypes.Secondary}>
      Next
      <Icon
        name={IconNames.arrow}
        decorative={true}
        iconRotation={IconRotationTypes.rotate270}
        modifiers={["small", "icon-right"]}
      />
    </Button>
  </div>
</Canvas>

Text in a button is optional if an icon is rendered, but make sure that there
is an `aria-label` passed to the `Button` component through the `attributes` prop:
`attributes={{ "aria-label": "Previous" }}`

<Canvas>
  <div
    style={{
      display: "flex",
      justifyContent: "space-between",
      width: "200px",
    }}
  >
    <Button
      buttonType={ButtonTypes.Secondary}
      attributes={{
        "aria-label": "Previous",
      }}
    >
      <Icon
        name={IconNames.arrow}
        decorative={true}
        iconRotation={IconRotationTypes.rotate90}
        modifiers={["small"]}
      />
    </Button>
    <Button
      buttonType={ButtonTypes.Secondary}
      attributes={{
        "aria-label": "Next",
      }}
    >
      <Icon
        name={IconNames.arrow}
        decorative={true}
        iconRotation={IconRotationTypes.rotate270}
        modifiers={["small"]}
      />
    </Button>
    <Button
      buttonType={ButtonTypes.Secondary}
      attributes={{
        "aria-label": "Close",
      }}
    >
      <Icon name={IconNames.close} decorative={true} modifiers={["small"]} />
    </Button>
  </div>
</Canvas>

## Patterns

Buttons are blue by default but should be red when it is intended as a call out.
The "Back to Top" button must be of `ButtonTypes` `Secondary` and must include an up arrow icon.

<Canvas>
  <Story name="Patterns">
    <Button
      buttonType={ButtonTypes.Primary}
      attributes={{
        style: {
          backgroundColor: getCSSVariable("--section-books-and-more-primary"),
        },
      }}
    >
      Donate to this library
    </Button>
    <br />
    <Button buttonType={ButtonTypes.Secondary}>
      Back to Top
      <Icon
        name={IconNames.arrow}
        decorative={true}
        iconRotation={IconRotationTypes.rotate180}
        modifiers={["small", "icon-right"]}
      />
    </Button>
  </Story>
</Canvas>

## Other Examples

Modifying the `buttonType` prop:

<Canvas>
  <Button buttonType={ButtonTypes.Primary}>Primary</Button>
  <Button buttonType={ButtonTypes.Secondary}>Secondary</Button>
  <Button buttonType={ButtonTypes.Pill}>Pill</Button>
  <Button buttonType={ButtonTypes.Link}>Link</Button>
</Canvas>

Modifying the `disabled` prop:

<Canvas>
  <Button disabled={false}>Enabled</Button>
  <Button disabled={true}>Disabled</Button>
</Canvas>
