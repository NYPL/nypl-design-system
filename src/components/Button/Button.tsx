import * as React from "react";
import bem from "../../utils/bem";
import { ButtonTypes } from "./ButtonTypes";

interface ButtonProps {
  /** Additional attributes passed to the button */
  attributes?: { [key: string]: any };
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** The kind of button assigned through the `ButtonTypes` enum  */
  buttonType?: ButtonTypes;
  /** Additional className for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  className?: string;
  /** Adds 'disabled' property to the button */
  disabled?: boolean;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Trigger the Button's action through the `mouseDown` event handler instead of `onClick`. `false` by default. */
  mouseDown?: boolean;
  /** The action to perform on the `<button>`'s onClick function */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /** The html button attribute */
  type?: "submit" | "button" | "reset";
}

/** Renders a simple `button` element with custom classes and modifiers. */
export default class Button extends React.Component<ButtonProps, any> {
  static defaultProps = {
    mouseDown: false,
  };

  constructor(props: ButtonProps) {
    super(props);
  }

  render(): JSX.Element {
    const {
      attributes,
      blockName,
      buttonType,
      children,
      className,
      disabled,
      id,
      modifiers = [],
      mouseDown,
      onClick,
      type = "submit",
    } = this.props;

    const buttonModifiers = modifiers ? modifiers : [];
    if (!buttonModifiers.find(modifier => modifier === buttonType)) {
      buttonModifiers.push(buttonType);
    }

    const baseClass = "button";

    const btnCallback = mouseDown
      ? { onMouseDown: onClick }
      : { onClick: onClick };

    return (
      <button
        id={id}
        className={`${baseClass} ${bem(baseClass, buttonModifiers, blockName, [
          className,
        ])}`}
        type={type}
        disabled={disabled}
        {...attributes}
        {...btnCallback}
      >
        {children}
      </button>
    );
  }
}
