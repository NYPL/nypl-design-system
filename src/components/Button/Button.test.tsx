import { expect } from "chai";
import { stub } from "sinon";
import * as Enzyme from "enzyme";
import * as React from "react";

import Button from "./Button";

describe("Button", () => {
  let wrapper: Enzyme.ShallowWrapper<any, any>;
  let onClick;

  beforeEach(() => {
    onClick = stub();
    wrapper = Enzyme.shallow(
      <Button id="button" onClick={onClick}>
        Submit
      </Button>
    );
  });

  it("calls the onClick", () => {
    wrapper.simulate("click");
    expect(onClick.callCount).to.equal(1);
  });
  it("optionally renders a component", () => {
    expect(wrapper.text()).to.equal("Submit");
  });
  it("has 'button' class", () => {
    expect(wrapper.find(".button")).to.have.lengthOf(1);
  });

  it("has type submit", () => {
    expect(wrapper.prop("type")).to.equal("submit");
  });

  it("optionally calls the onClick on mouseDown instead of on click", () => {
    expect(onClick.callCount).to.equal(0);
    wrapper.setProps({ mouseDown: true });
    wrapper.simulate("click");
    expect(onClick.callCount).to.equal(0);
    wrapper.simulate("mouseDown");
    expect(onClick.callCount).to.equal(1);
  });

  describe("rendering content from its children prop", () => {
    it("should render string children", () => {
      wrapper = Enzyme.shallow(
        <Button id="button" onClick={onClick}>
          Submit
        </Button>
      );
      expect(wrapper.text()).to.equal("Submit");
    });

    it("should render element children", () => {
      wrapper = Enzyme.shallow(
        <Button id="button" onClick={onClick}>
          <em>I'm a em element</em>
        </Button>
      );
      expect(wrapper.text()).to.equal("I'm a em element");
      expect(wrapper.find("em")).to.have.lengthOf(1);
    });
  });

  describe("passes button type", () => {
    it("button has passed type", () => {
      const wrapper = Enzyme.mount(
        <Button id="button" onClick={onClick} type="button">
          <span>{"I'm a span element"}</span>
        </Button>
      );
      expect(wrapper.find("button").prop("type")).to.equal("button");
    });
  });
});
