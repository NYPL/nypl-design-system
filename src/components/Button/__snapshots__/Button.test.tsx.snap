// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot Renders the UI snapshot correctly 1`] = `
<button
  className="button button button--undefined "
  id="button"
  onClick={[MockFunction]}
  type="submit"
>
  Submit
</button>
`;
